File: next/dweb-browser/MicroService/src/Sys/Boot/BootNMM.cs

namespace DwebBrowser.MicroService.Sys.Boot;

public class BootNMM : NativeMicroModule
{
    static readonly Debugger Console = new("BootNMM");
    /**
     * 开机启动项注册表
     * TODO 这里需要从数据库中读取
     */
    private HashSet<Mmid> _registeredMmids = new();
    public override List<MicroModuleCategory> Categories { get; init; } = new()
    {
        MicroModuleCategory.Service,
        MicroModuleCategory.Hub_Service,
    };

    public override string? ShortName { get; set; } = "Boot";
    public BootNMM(List<Mmid>? initMmids = null) : base("boot.sys.dweb", "Boot Management")
    {
        if (initMmids is not null)
        {
            _registeredMmids.UnionWith(initMmids.ToHashSet());
        }
    }

    protected override async Task _bootstrapAsync(IBootstrapContext bootstrapContext)
    {
        HttpRouter.AddRoute(IpcMethod.Get, "/register", async (_, ipc) =>
        {
            return _register(ipc!.Remote.Mmid);
        });
        HttpRouter.AddRoute(IpcMethod.Get, "/unregister", async (_, ipc) =>
        {
            return _unregister(ipc!.Remote.Mmid);
        });

        /// 基于activity事件来启动开机项
        OnActivity += async (Event, ipc, _) =>
        {
            // 只响应 dns 模块的激活事件
            if (ipc.Remote.Mmid != "dns.std.dweb")
            {
                return;
            }

            foreach (var mmid in _registeredMmids)
            {
                Console.Log("OnActivity", "launch {0}", mmid);
                var fromIpc = await ConnectAsync(mmid);
                await (fromIpc?.PostMessageAsync(Event)).ForAwait();
            }
        };
    }

    protected override async Task _shutdownAsync()
    {
        Router!.Clear();
    }

    /**
     * <summary>
     * 注册一个boot程序
     * TODO 这里应该有用户授权，允许开机启动
     * <summary>
     */
    private bool _register(Mmid mmid) => _registeredMmids.Add(mmid);

    /**
     * <summary>
     * 移除一个boot程序
     * TODO 这里应该有用户授权，取消开机启动
     * </summary>
     */
    private bool _unregister(Mmid mmid) => _registeredMmids.Remove(mmid);
}

File: next/kmp/core/src/commonMain/kotlin/org/dweb_browser/core/std/boot/BootNMM.kt
package org.dweb_browser.core.std.boot

import org.dweb_browser.core.help.types.MICRO_MODULE_CATEGORY
import org.dweb_browser.core.help.types.MMID
import org.dweb_browser.core.http.router.bind
import org.dweb_browser.core.module.BootstrapContext
import org.dweb_browser.core.module.NativeMicroModule
import org.dweb_browser.core.module.Router
import org.dweb_browser.core.std.dns.ext.onActivity
import org.dweb_browser.helper.Debugger
import org.dweb_browser.pure.http.PureMethod

val debugBoot = Debugger("boot")

class BootNMM(val initMmids: List<MMID>? = null) :
  NativeMicroModule("boot.sys.dweb", "Boot Management") {
  init {
    short_name = "Boot";
    categories = mutableListOf(MICRO_MODULE_CATEGORY.Service, MICRO_MODULE_CATEGORY.Hub_Service)
  }

  inner class BootRuntime(override val bootstrapContext: BootstrapContext) : NativeRuntime() {

    /**
     * 开机启动项注册表
     * TODO 这里需要从数据库中读取
     */
    private val registeredMmids = mutableSetOf<MMID>()

    init {
      if (initMmids != null) {
        registeredMmids += initMmids
      }
    }

    override val routers: Router = mutableMapOf()
    override suspend fun _bootstrap() {
      routes("/register" bind PureMethod.GET by defineBooleanResponse {
        register(ipc.remote.mmid)
      }, "/unregister" bind PureMethod.GET by defineBooleanResponse {
        unregister(ipc.remote.mmid)
      }).protected("dns.std.dweb")
      for (mmid in registeredMmids) {
        debugBoot("open", mmid)
        bootstrapContext.dns.open(mmid)
      }
      onActivity { (event, ipc) ->
        // 只响应 dns 模块的激活事件
        if (ipc.remote.mmid == "dns.std.dweb") {
          for (mmid in registeredMmids) {
            debugBoot("activity", mmid)
            connect(mmid).postMessage(event)
          }
        }
      }
    }


    override suspend fun _shutdown() {
      routers.clear()
    }

    /**
     * 注册一个boot程序
     * TODO 这里应该有用户授权，允许开机启动
     */
    private fun register(mmid: MMID) = this.registeredMmids.add(mmid);

    /**
     * 移除一个boot程序
     * TODO 这里应该有用户授权，取消开机启动
     */
    private fun unregister(mmid: MMID) = this.registeredMmids.remove(mmid);

  }

  override fun createRuntime(bootstrapContext: BootstrapContext) = BootRuntime(bootstrapContext)
}
